### PR TITLE
Simplify ExprUseVisitor

### PR DESCRIPTION
**Pull request by matthewjasper**
* Remove HIR const qualification
* Remove parts of ExprUseVisitor that aren't being used

r? @eddyb

### CODE REVIEW

**Comment by eddyb** [comment]
I just noticed the old `regionck` is still around - do we have a timeline for removing it?

File: src/librustc/middle/mem_categorization.rs
Comment (eddyb) [inline src/librustc/middle/mem_categorization.rs]: Can that lifetime be removed from MC? It would be wrong now, so I hope nothing uses it.

File: src/librustc/middle/expr_use_visitor.rs
Comment (oli-obk) [inline src/librustc/middle/expr_use_visitor.rs]: can you check whether clippy still works after your PR? We use this visitor a lot.

File: src/test/ui/issues/issue-23302-3.stderr
Comment (eddyb) [inline src/test/ui/issues/issue-23302-3.stderr]: Hmm, I think maybe `mir_const_qualif` or whatever it's called should have a message along the lines of `... const-checking A`.

File: src/librustc_passes/lib.rs
Comment (eddyb) [inline src/librustc_passes/lib.rs]: Feels like this crate is becoming quite pointless.

File: src/librustc_typeck/check/upvar.rs
Comment (eddyb) [inline src/librustc_typeck/check/upvar.rs]: Is this the only user left?

File: src/librustc/middle/expr_use_visitor.rs
Comment (eddyb) [reply src/librustc/middle/expr_use_visitor.rs]: Oh... anything using EUV should've transitioned a long time ago to being based on MIR, IMO. I wish I brought this up to clippy people earlier :/

**Comment by matthewjasper** [comment]
The old `regionck` is still used for items. A lot of it can be removed now, but that's another PR.

Notes for anyone trying to update Clippy after this lands:

* `utils/usage.rs` - doesn't use anything removed here
* `loops.rs` - needs to change from the span passed in to the span of the `cmt`.
* `escape.rs` - `consume` and `consume_pat` will need to be merged
* `needless_pass_by_value` - should probably use `borrow` instead of `matched_pat`

File: src/librustc_passes/lib.rs
Comment (Centril) [reply src/librustc_passes/lib.rs]: We could split them out to a crate for each pass. :)

File: src/librustc/middle/mem_categorization.rs
Comment (matthewjasper) [reply src/librustc/middle/mem_categorization.rs]: `regionck` was using it, but it was already getting the wrong lifetimes.

File: src/librustc_passes/lib.rs
Comment (eddyb) [reply src/librustc_passes/lib.rs]: They are tiny though... a bunch of them belong in `rustc::hir` which *could* become a crate (with only the data types remaining in `rustc`).

Comment (Centril) [reply src/librustc_passes/lib.rs]: Whatever we do, let's please not inflate the size of the `rustc` crate. :)

**Review by eddyb — APPROVED** [review]
LGTM, r=me if you don't want to change more due to https://github.com/rust-lang/rust/pull/64874#discussion_r329314823

**Review by nikomatsakis — APPROVED** [review]
Nice cleanup, @matthewjasper!

**Pull request merged**
